File: wlr/OGUR/OGUR/HUD/InventoryHud.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;
using OGUR.Creatures;
using OGUR.Items;
using SPX.Util;
using SPX.Core;
using OGUR.Management;

namespace OGUR.HUD
{
    public class InventoryHud:IHud
    {
        private int m_currentClass = ItemClass.Values[1];
        private readonly Inventory m_inventory;
        private readonly Equipment m_equipment;
        private int m_endingItem = 4, m_startingItem = 0;
        private Dictionary<GenericItem, int> m_currentClassItems;
        private GenericItem m_currentSelectedItem = null;

        private EquipmentHud m_equipHud;
        private DeltasHud m_deltas;

        public InventoryHud(ICreature owner,Inventory inventory,Equipment equipment):base(owner,XnaManager.WindowWidth/2,XnaManager.WindowHeight/2)
        {
            m_inventory = inventory;
            m_equipment = equipment;
            m_deltas = new DeltasHud(owner,equipment);
            m_equipHud = new EquipmentHud(owner, equipment);
            m_currentClassItems = m_inventory.GetItems(m_currentClass);
        }

        public void Draw()
        {
            if (m_isVisible)
            {
                XnaManager.Renderer.Draw(m_menuBase,GetHudOrigin(), new Rectangle(0, 0, 1, 1), Color.White, 0f, new Vector2(0,0), XnaManager.GetCenter(), SpriteEffects.None,ZDepth.HudBG);
                m_textHandler.Draw();
                m_deltas.Draw();
                m_equipHud.Draw();
            }
        }

        private void HandleInput()
        {
            if (Input.IsPressed(Commands.CycleLeft, m_parent.GetPlayerIndex()))
            {
                m_currentClass--;
                if (m_currentClass <= ItemClass.NULL)
                {
                    m_currentClass = ItemClass.Values.Count() - 2;
                }
                m_startingItem = 0;
                m_endingItem = 4;
                forceRefresh = true;
            }

            if (Input.IsPressed(Commands.CycleRight, m_parent.GetPlayerIndex()))
            {
                m_currentClass++;
                if (m_currentClass >= ItemClass.LAST)
                {
                    m_currentClass = ItemClass.Values[1];                    
                }
                m_startingItem = 0;
                m_endingItem = 4;
                forceRefresh = true;
            }

            if (Input.IsPressed(Commands.MoveDown, m_parent.GetPlayerIndex()))
            {
                if (m_startingItem < m_currentClassItems.Count() - 1)
                {
                    m_startingItem++;
                    m_endingItem++;
                    forceRefresh = true;
                }
            }

            if (Input.IsPressed(Commands.MoveUp, m_parent.GetPlayerIndex()))
            {
                if (m_startingItem > 0)
                {
                    m_startingItem--;
                    m_endingItem--;
                    forceRefresh = true;
                }
            }
            if (Input.IsPressed(Commands.Confirm, m_parent.GetPlayerIndex()))
            {
                m_parent.Equip(m_currentSelectedItem);
                forceRefresh = true;
            }
            if (Input.IsPressed(Commands.Cancel, m_parent.GetPlayerIndex()))
            {
                m_parent.Drop(m_currentSelectedItem);
                forceRefresh = true;
            }
        }
        public void Update()
        {
            if (m_isVisible)
            {
                HandleInput();
                m_textHandler.Update();
                m_deltas.Update(m_currentSelectedItem,forceRefresh);
                m_equipHud.Update(forceRefresh);
                m_textHandler.Clear();
                UpdateInventoryDisplay();
                if (forceRefresh)
                {
                    forceRefresh = false;
                }
            }
        }

        public override void Toggle()
        {
            base.Toggle();
            m_deltas.Toggle();
            m_equipHud.Toggle();
            forceRefresh = true;
        }

        private static readonly Dictionary<int, string> s_classStrings = new Dictionary<int, string>();

        private string GetClassDisplay()
        {
            if(!s_classStrings.ContainsKey(m_currentClass))
            {
                s_classStrings.Add(m_currentClass, ItemClass.Names[m_currentClass]);
            }
            return s_classStrings[m_currentClass];
        }

        private const string s_delim = ")";
        private const string s_equipDelim = "~";
        private const string s_seper = " x";
        private const string s_newline = "\n";

        private string displayString = "";
        private bool forceRefresh = false;

        private void UpdateInventoryDisplay()
        {
            m_textHandler.WriteDefault(GetClassDisplay(), 20, 30,GetHudOrigin());
            m_currentClassItems = m_inventory.GetItems(m_currentClass);
            if (m_currentClassItems.Count > 0)
            {
                int ii = 0;
                if (forceRefresh)
                {
                    StringSquisher.Clear();
                    foreach (var item in m_currentClassItems.Keys)
                    {
                        if (ii == m_startingItem)
                        {
                            m_currentSelectedItem = item;
                        }
                        if(!m_equipment.IsRegistered(item)&&m_inventory.GetItemCount(item)<=0)
                        {
                            continue;
                        }

                   
                        if (ii >= m_startingItem && ii < m_endingItem && ii < m_currentClassItems.Keys.Count())
                        {
                            StringSquisher.Squish
                            (
                                StringStorage.Get(ii), s_delim,
                                (m_equipment.IsRegistered(item)) ? s_equipDelim : String.Empty,
                                item.Name);
                            if (m_currentClassItems[item] > -1)
                            {
                                StringSquisher.Squish(s_seper, StringStorage.Get(m_currentClassItems[item]));
                            }
                            StringSquisher.Squish(s_newline);
                        }
                        ii++;
                    }
                    displayString = StringSquisher.Flush();       
                }
                m_textHandler.WriteDefault(displayString, 50, 60, GetHudOrigin());
            }
        }
    }
}

File: proj/src/aigilas/hud/InventoryHud.java
package aigilas.hud;

import java.util.HashMap;

import spx.bridge.DrawDepth;
import spx.core.SpxManager;
import spx.io.Input;
import spx.util.StringSquisher;
import spx.util.StringStorage;
import aigilas.creatures.ICreature;
import aigilas.items.Equipment;
import aigilas.items.GenericItem;
import aigilas.items.Inventory;
import aigilas.items.ItemClass;
import aigilas.management.Commands;

import com.badlogic.gdx.graphics.Color;

public class InventoryHud extends IHud {
	private int _currentClass = 0;
	private final Inventory _inventory;
	private final Equipment _equipment;
	private int _endingItem = 4;
	private int _startingItem = 0;
	private HashMap<GenericItem, Integer> _currentClassItems;
	private GenericItem _currentSelectedItem = null;

	private final EquipmentHud _equipHud;
	private final DeltasHud _deltas;

	public InventoryHud(ICreature owner, Inventory inventory, Equipment equipment) {
		super(owner, SpxManager.WindowWidth / 2, SpxManager.WindowHeight / 2);
		_inventory = inventory;
		_equipment = equipment;
		_deltas = new DeltasHud(owner, equipment);
		_equipHud = new EquipmentHud(owner, equipment);
		_currentClassItems = _inventory.GetItems(ItemClass.values()[_currentClass]);
	}

	public void Draw() {
		if (_isVisible) {
			SpxManager.Renderer.Draw(_menuBase, GetHudOrigin(), DrawDepth.HudBG, Color.BLACK, (int) SpxManager.GetCenter().X, (int) SpxManager.GetCenter().Y);
			_textHandler.Draw();
			_deltas.Draw();
			_equipHud.Draw();
		}
	}

	private void HandleInput() {
		if (Input.IsActive(Commands.CycleLeft, _parent.GetPlayerIndex())) {
			_currentClass--;
			if (_currentClass <= 0) {
				_currentClass = ItemClass.values().length - 2;
			}
			_startingItem = 0;
			_endingItem = 4;
			forceRefresh = true;
		}

		if (Input.IsActive(Commands.CycleRight, _parent.GetPlayerIndex())) {
			_currentClass++;
			if (_currentClass >= ItemClass.values().length) {
				_currentClass = 1;
			}
			_startingItem = 0;
			_endingItem = 4;
			forceRefresh = true;
		}

		if (Input.IsActive(Commands.MoveDown, _parent.GetPlayerIndex())) {
			if (_startingItem < _currentClassItems.size() - 1) {
				_startingItem++;
				_endingItem++;
				forceRefresh = true;
			}
		}

		if (Input.IsActive(Commands.MoveUp, _parent.GetPlayerIndex())) {
			if (_startingItem > 0) {
				_startingItem--;
				_endingItem--;
				forceRefresh = true;
			}
		}
		if (Input.IsActive(Commands.Confirm, _parent.GetPlayerIndex())) {
			_parent.Equip(_currentSelectedItem);
			forceRefresh = true;
		}
		if (Input.IsActive(Commands.Cancel, _parent.GetPlayerIndex())) {
			_parent.Drop(_currentSelectedItem);
			forceRefresh = true;
		}
	}

	public void Update() {
		if (_isVisible) {
			HandleInput();
			_textHandler.Update();
			_deltas.Update(_currentSelectedItem, forceRefresh);
			_equipHud.Update(forceRefresh);
			_textHandler.Clear();
			UpdateInventoryDisplay();
			if (forceRefresh) {
				forceRefresh = false;
			}
		}
	}

	@Override
	public void Toggle() {
		super.Toggle();
		_deltas.Toggle();
		_equipHud.Toggle();
		forceRefresh = true;
	}

	private static HashMap<Integer, String> __classStrings = new HashMap<>();

	private String GetClassDisplay() {
		if (!__classStrings.containsKey(_currentClass)) {
			__classStrings.put(_currentClass, ItemClass.values()[_currentClass].hudName());
		}
		return __classStrings.get(_currentClass);
	}

	private static final String __delim = ")";
	private static final String __equipDelim = "~";
	private static final String __seper = " x";
	private static final String __newline = "\n";

	private String displayString = "";
	private boolean forceRefresh = false;
	private String[] list = new String[10];

	private void UpdateInventoryDisplay() {
		_textHandler.WriteDefault(GetClassDisplay(), 20, 30, GetHudOrigin());
		_currentClassItems = _inventory.GetItems(ItemClass.values()[_currentClass]);
		if (_currentClassItems.size() > 0) {
			int ii = 0;
			if (forceRefresh) {
				StringSquisher.Clear();
				displayString = StringSquisher.Flush();
				int count = 0;
				for (GenericItem item : _currentClassItems.keySet()) {
					if (ii == _startingItem) {
						_currentSelectedItem = item;
					}
					if (!_equipment.IsRegistered(item) && _inventory.GetItemCount(item) <= 0) {
						continue;
					}

					if (ii >= _startingItem && ii < _endingItem && ii < _currentClassItems.keySet().size()) {
						count++;
						StringSquisher.Clear();
						StringSquisher.Squish(StringStorage.Get(ii), __delim, (_equipment.IsRegistered(item)) ? __equipDelim : "", item.Name);
						if (_currentClassItems.get(item) > -1) {
							StringSquisher.Squish(__seper, StringStorage.Get(_currentClassItems.get(item)));
						}
						StringSquisher.Squish(__newline);
						if (count < 10) {
							list[count] = StringSquisher.Flush();
						}
					}
					ii++;
				}

			}
			_textHandler.WriteDefault(displayString, 50, 60, GetHudOrigin());
			for (int jj = 0; jj < 10; jj++) {
				if (list[jj] != null) {
					_textHandler.WriteDefault(list[jj], 20, 60 + 40 * jj, GetHudOrigin());
				}
			}
		}
	}
}
